fix: prevent policies targeting non-TLS listeners on the same port from conflicting

liorokman opened the pull request:
**What this PR does / why we need it**:

* Validate that multiple policies that affect listener configuration don't map to the same listener filter chain.
* Change the XDS listener generation so that instead of defaultFilterChain for non-TLS routes, a filterChain with a destinationPort matcher is used. This allows multiple policies attached to non-TLS listeners that differ on the destination port to provide different policies without conflicting.

**Which issue(s) this PR fixes**:
Fixes #2742

arkodg commented:
thanks for working on this @liorokman, since this is a pretty big change in terms of how xds is generated, this is unlikely to make it into v1.0.0, will take a closer look at the PR soon

guydc commented on internal/xds/translator/listener.go:
```diff
@@ -293,18 +293,32 @@ func (t *Translator) addXdsHTTPFilterChain(xdsListener *listenerv3.Listener, irL
+		filterChain.FilterChainMatch = &listenerv3.FilterChainMatch{
+			DestinationPort: wrapperspb.UInt32(irListener.Port),
 		}
-		xdsListener.DefaultFilterChain = filterChain
```
raised in yesterday's community meeting: when matching on the destination port, wouldn't there be a dedicated envoy listener that's already bound to that distinct port? meaning, for port-based matching, the default filter chain can still be used and there's no need to additionally do a filter chain match?

arkodg commented:
hey @liorokman, @guydc, @zhaohuabing and I discussed this issue in the community meeting yesterday, and prefer if we just partially rejected/skipped the policy if we detected this condition w/o changing any xds translation. 
It may be easier to implement first policy wins, and partially reject the second policy for this listener (similar to the `overlap` condition)
I had also raised this issue in the envoy slack channel, but didn't find a solution 
https://envoyproxy.slack.com/archives/C78M4KW76/p1709687283123669

liorokman commented:
It's not possible to reject or partially accept the second policy in this case because the problem is in the XDS translation and not the IR translation. The IR translation doesn't have an issue - when examining the IR for the listener that is _actually_ targeted by a policy there is no conflict. The conflict occurs because the XDS translation mashes multiple IRs into the same `defaultFilterContext`. 

In other words - it's not that two policies are in conflict with each other. It's that one single policy is asked to target a specific Listener and ends up affecting more than that.

Instead, the PR right now rejects policies that would affect listeners that they weren't the target of.

I made the XDS translation change because that way I could accept _more_ policies. Instead of declining to accept all policies targeting non-TLS listeners and routes attached to non-TLS listeners, the PR right now makes it possible to accept policies that don't target the same port.

liorokman commented:
Based on @guydc 's insight - no need to change the `defaultFilterChain` translation since a new listener is created anyways per port. I reverted that part of the patch.

What remains is that if a policy (either SecurityPolicy or ClientTrafficPolicy) is applied that would affect more than the listener that it was attached to, then the policy will be rejected.

arkodg commented:
@liorokman there is existing logic there checks for conflicts using a policyMap https://github.com/envoyproxy/gateway/blob/e4aea1b6e51f1a3810345e34cba4e39a94bbb5bb/internal/gatewayapi/clienttrafficpolicy.go#L72

do you think we should reuse that style, and add a `httpOnlyMap` that can make sure no 2 policies can affect the same http only port ?

I dont think we need to include `mergeGateways` since that feature has already kicked in earlier in the routing level and flattened all gateway listeners into IR listeners associated with a single key (per gatewayclass)

liorokman commented:
> I dont think we need to include `mergeGateways` since that feature has already kicked in earlier in the routing level and flattened all gateway listeners into IR listeners associated with a single key (per gatewayclass)


Good point. Fixed.

liorokman commented:
> @liorokman there is existing logic there checks for conflicts using a policyMap
> do you think we should reuse that style, and add a httpOnlyMap that can make sure no 2 policies can affect the same http only port ?

I tried to do this, and it didn't end up being any simpler to read. Is this critical?

arkodg commented:
> > @liorokman there is existing logic there checks for conflicts using a policyMap
> > do you think we should reuse that style, and add a httpOnlyMap that can make sure no 2 policies can affect the same http only port ?
> 
> I tried to do this, and it didn't end up being any simpler to read. Is this critical?

it helps keep the code and corresponding status (`Accepted=False` with `Conflicted` Reason) in the same place.
The current code checks that 2 policies cannot target the same `section` i.e. listener, this case is a variant where 2 policies cannot target the same `http` port

liorokman commented:
> > > @liorokman there is existing logic there checks for conflicts using a policyMap
> > > do you think we should reuse that style, and add a httpOnlyMap that can make sure no 2 policies can affect the same http only port ?
> > 
> > 
> > I tried to do this, and it didn't end up being any simpler to read. Is this critical?
> 
> it helps keep the code and corresponding status (`Accepted=False` with `Conflicted` Reason) in the same place. The current code checks that 2 policies cannot target the same `section` i.e. listener, this case is a variant where 2 policies cannot target the same `http` port

The current code that checks that 2 policies cannot target the same `section` doesn't have to look at information from inside the listener(s) identified by that section. It can do its job by looking at the strings in the policy itself.

To determine the port(s) targeted by a single policy document, the code needs to find the actual listener, or multiple listeners in the case where the policy attaches to an entire `Gateway`. We actually already have code that does this - the translation functions where the current PR does the validation.

Moving the check from the place where all of the information to do the validation exists to where this information doesn't yet exist would more or less duplicate the code from the translation functions. Is this your preference?

arkodg commented:
> Moving the check from the place where all of the information to do the validation exists to where this information doesn't yet exist would more or less duplicate the code from the translation functions. Is this your preference?

hoping others can chime in here as well, my preference would be to separate translation from validation, so the method could be called after getting the IR context before calling translate

also would be great if this layer did not have any envoyisms in the gateway-api layer, although we are in this situation because of the limitation in Envoy proxy, would be great to highlight that as a code comment, but use a more generic name (instead of `ShareEnvoyFilterChain`) for the helper method

liorokman commented:
> > Moving the check from the place where all of the information to do the validation exists to where this information doesn't yet exist would more or less duplicate the code from the translation functions. Is this your preference?
> 
> hoping others can chime in here as well, my preference would be to separate translation from validation, so the method could be called after getting the IR context before calling translate
> 
> also would be great if this layer did not have any envoyisms in the gateway-api layer, although we are in this situation because of the limitation in Envoy proxy, would be great to highlight that as a code comment, but use a more generic name (instead of `ShareEnvoyFilterChain`) for the helper method

I split the validation away from the translation, and renamed the `ShareEnvoyFilterChain` function to something a little more generic.

arkodg commented on internal/gatewayapi/clienttrafficpolicy.go:
```diff
@@ -285,6 +302,29 @@ func resolveCTPolicyTargetRef(policy *egv1a1.ClientTrafficPolicy, gateways []*Ga
 	return gateway
 }
 
+func validatePortOverlapForClientTrafficPolicy(l *ListenerContext, xds *ir.Xds) error {
```
can this be also shared with SecurityPolicy and moved into helpers.go ?

liorokman replied:
Not easily. CTP is listener oriented, while SP is route oriented. The way that the correct listener is found is different in each of these validation methods.

arkodg reviewed (dismissed):
LGTM thanks !
